### PR TITLE
Signup

### PR DESCRIPTION
**Pull request by ivanfelfo**
signup feature added

### CODE REVIEW

File: app/controllers/v1/registrations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class V1::RegistrationsController < ApplicationController
+    
+
```
Comment (martinjaimem) [inline app/controllers/v1/registrations_controller.rb:3]: You are leaving some blank spaces here, what about using `rubocop`?

File: db/migrate/20210415121835_add_gender_column.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class AddGenderColumn < ActiveRecord::Migration[6.1]
+  def up
+    add_column(:users, :gender, :boolean)
```
Comment (martinjaimem) [inline db/migrate/20210415121835_add_gender_column.rb:3]: Do you have any specific reason about why using a boolean field here?

File: log/development.log [hunk 3]
```diff
@@ -164,3 +164,360 @@ Processing by Rails::WelcomeController#index as HTML
 Completed 200 OK in 28ms (Views: 13.1ms | ActiveRecord: 0.0ms | Allocations: 2362)
 
 
+DEPRECATION WARNING: [] is deprecated and will be removed from Rails 6.2 (Use configs_for) (called from migration_template at (erb):6)
```
Comment (martinjaimem) [inline log/development.log:167]: This file should not be committed. In fact, I think you added it before changing the `.gitignore`

File: app/controllers/v1/registrations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class V1::RegistrationsController < ApplicationController
+    
+    
+    def create
+        user = User.new(email: params[:email], gender: params[:gender] ,password: params[:password])
```
Comment (colembolos) [inline app/controllers/v1/registrations_controller.rb:5]: To avoid add attributes by one over here you can create a `strong param` and keep it simple like `User.new(user_params)`

File: app/controllers/v1/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+class V1::SessionsController < ApplicationController
+    def create
+        user = User.where(email: params[:email]).first
```
Comment (colembolos) [inline app/controllers/v1/sessions_controller.rb:3]: You can use `find_by` instead of where and avoid that `.first` because you want to find a record not a collection

File: app/controllers/v1/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+        if user && user.valid_password?(params[:password])
```
Comment (colembolos) [inline app/controllers/v1/sessions_controller.rb:5]: I'm not seeing something that it's created over here, what is the purpose of the controller?

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+end
```
Comment (colembolos) [inline app/models/user.rb:6]: You can add some validation for presence of the email and other attributes that you use

**Review by colembolos — COMMENTED** [review]
Good first step! I left some comments

File: config/routes.rb [hunk 8]
```diff
@@ -1,3 +1,10 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  devise_for :users
+
+  namespace :v1 do
+    resources :registrations, only: [:create]
+    resources :sessions, only: [:create]
+  end
```
Comment (colembolos) [inline config/routes.rb:7]: If you are using devise you don't need to create a custom route. Check this to configure the routes better https://nicedoc.io/heartcombo/devise

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+capybara-*.html
+/log
+/tmp
+/db/*.sqlite3
+/public/system
+/coverage/
+/spec/tmp
+/storage
+.idea/
+.DS_Store
```
Comment (juanmferreira93) [inline .gitignore:13]: you should add this file to the .gitignore too: `config/.DS_Store`

File: config/locales/devise.en.yml [hunk 10]
```diff
@@ -0,0 +1,65 @@
+# Additional translations at https://github.com/heartcombo/devise/wiki/I18n
+
+en:
+  devise:
+    confirmations:
+      confirmed: "Your email address has been successfully confirmed."
```
Comment (juanmferreira93) [inline config/locales/devise.en.yml:6]: `"` are not required here, you can just use: `Your email address has been successfully confirmed.` We could discuss later how to change all this `"` in only one action, don't mind for this now

File: db/migrate/20210415121835_add_gender_column.rb (hunk 2 above)
Comment (juanmferreira93) [reply db/migrate/20210415121835_add_gender_column.rb:3]: As I suggested before, we can use an `enum` here: https://api.rubyonrails.org/v5.2.5/classes/ActiveRecord/Enum.html

File: .gitignore [hunk 11]
```diff
@@ -0,0 +1,39 @@
+/config/database.yml
+*.rbc
+*.swp
+capybara-*.html
+/log
+/tmp
```
Comment (juanmferreira93) [inline .gitignore:6]: for some reason you have some `temp` files that are being tracked and avoid the .gitignore, we could see this later, don't worry

**Review by juanmferreira93 — COMMENTED** [review]
nice work 👏 , leave some comment and waiting for the changes we talked before

File: spec/requests/users/create_spec.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+RSpec.describe 'POST /v1/users', type: :request do
+  let(:params) { { user: { email: 'mail@mail.com', password: 'hello123', gender: true } } }
+  it 'saves user in db' do
+    post v1_user_registration_path, params: params, as: :json
+    user = User.last
+    expect(user.email).to eq('mail@mail.com')
+    expect(user.gender).to eq(true)
+  end
+
```
Comment (martinjaimem) [inline spec/requests/users/create_spec.rb:13]: what about checking the response status?

```
context 'when not valid' do
  it { is_expected.to respond_with 422 }
end
```

and when it is valid:


```
context 'when valid' do
  it { is_expected.to respond_with 200 }
end
```

File: spec/models/user_spec.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (martinjaimem) [inline spec/models/user_spec.rb:5]: Maybe we can do a separate pull request that adds a factory for the user and we check the validations:
https://www.betterspecs.org/#factories
what do you think?

File: .rspec [hunk 14]
```diff
@@ -0,0 +1 @@
+--require spec_helper
```
Comment (martinjaimem) [inline .rspec:1]:
```suggestion
--require rails_helper
```

File: app/controllers/v1/registrations_controller.rb [hunk 15]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module V1
+  class RegistrationsController < ApplicationController
+    rescue_from ActiveRecord::RecordInvalid, with: :render_record_invalid
```
Comment (martinjaimem) [inline app/controllers/v1/registrations_controller.rb:5]: Move this line and the method to the application controller

File: spec/requests/users/create_spec.rb [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+  context 'when trying to register a user with an email that already exists' do
+    let!(:created_user) { User.create(email: 'mail@mail.com', password: 'hello123', gender: true) }
+    it 'does not create a new user' do
```
Comment (martinjaimem) [inline spec/requests/users/create_spec.rb:16]: leave a blank line between `lets` and `its`

File: app/controllers/v1/sessions_controller.rb [hunk 17]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module V1
+  class SessionsController < ApplicationController
+    def create
+      user = User.find_by(email: params[:email])
```
Comment (martinjaimem) [inline app/controllers/v1/sessions_controller.rb:6]: are we sure this is working?
Can we add a test for this?

File: app/models/user.rb [hunk 18]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable, :confirmable
+
+  validates :email, :presence => true, :uniqueness =>  true
```
[the remark this answers is not in the document]
Comment (ivanfelfo) [reply app/models/user.rb]: 👍

File: .gitignore (hunk 9 above)
Comment (ivanfelfo) [reply .gitignore:13]: 👍

File: spec/requests/users/create_spec.rb (hunk 12 above)
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:13]: 👍

File: spec/models/user_spec.rb (hunk 13 above)
Comment (ivanfelfo) [reply spec/models/user_spec.rb:5]: i think thats a great idea! :)

File: .gitignore [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+/vendor/bundle
+
+# these should all be checked in to normalise the environment:
+# Gemfile.lock, .ruby-version, .ruby-gemset
+
+# unless supporting rvm < 1.11.0 or doing something fancy, ignore this:
+.rvmrc
+
+# Ignore byebug history
+.byebug_history
```
Comment (juanmferreira93) [inline .gitignore:33]: probably we will have to manually delete this file for the .gitignore to start working with it

File: spec/requests/sessions/create_spec.rb [hunk 20]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+RSpec.describe 'POST /v1/users/sign_in', type: :request do
+  subject { post v1_user_session_path, params: params, as: :json }
+  let(:params) { { user: { email: 'mail@mail.com', password: 'hello123' } } }
+
+  context 'when user is found at db' do
+    let!(:created_user) { User.create!(email: 'mail@mail.com', password: 'hello123') }
```
Comment (juanmferreira93) [inline spec/requests/sessions/create_spec.rb:9]: We will discuss later about another way to create fake models to test. Probably we could set up it in another PR. That's work for now 🚀

File: spec/requests/sessions/create_spec.rb [hunk 21]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+RSpec.describe 'POST /v1/users/sign_in', type: :request do
+  subject { post v1_user_session_path, params: params, as: :json }
+  let(:params) { { user: { email: 'mail@mail.com', password: 'hello123' } } }
+
+  context 'when user is found at db' do
```
Comment (juanmferreira93) [inline spec/requests/sessions/create_spec.rb:8]: maybe, `when the user already exists`, wdyt? Same for others

File: spec/requests/sessions/create_spec.rb [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+    let!(:created_user) { User.create!(email: 'mail@mail.com', password: 'hello123') }
+
+    context 'when password matches' do
+      it 'returns code 200' do
```
Comment (juanmferreira93) [inline spec/requests/sessions/create_spec.rb:12]: `responds with code 200` maybe better, wdyt? Same for the similar ones

File: spec/requests/sessions/create_spec.rb [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+        expect(response).to have_http_status(201)
+      end
+
+      it 'returns user email' do
+        subject
+        expect(JSON.parse(response.body)['email']).to eq(created_user.email)
+      end
+    end
+
+    context 'when password does not match' do
```
Comment (juanmferreira93) [inline spec/requests/sessions/create_spec.rb:23]: what's happens if we leave one param blank? Like mail or password?

File: spec/requests/users/create_spec.rb [hunk 24]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'POST /v1/users', type: :request do
+  subject { post v1_user_registration_path, params: params, as: :json }
+  let(:params) { { user: { email: 'mail@mail.com', password: 'hello123', gender: true } } }
+
+  context 'when valid' do
+    it 'saves user in db' do
```
Comment (juanmferreira93) [inline spec/requests/users/create_spec.rb:10]: `creates an user` maybe

File: spec/requests/users/create_spec.rb [hunk 25]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'POST /v1/users', type: :request do
+  subject { post v1_user_registration_path, params: params, as: :json }
+  let(:params) { { user: { email: 'mail@mail.com', password: 'hello123', gender: true } } }
```
Comment (juanmferreira93) [inline spec/requests/users/create_spec.rb:7]: what's mean 'gender: true'? Ah, I think that the enhancement for this is in another PR, 👌

File: spec/requests/users/create_spec.rb [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+
+  context 'when valid' do
+    it 'saves user in db' do
+      subject
+      user = User.last
+      expect(user.email).to eq('mail@mail.com')
+      expect(user.gender).to eq(true)
+    end
+
```
Comment (juanmferreira93) [inline spec/requests/users/create_spec.rb:16]: we could expect that the `count` of Users would be changed, something like: `expect { subject }.to change(User, :count).by(1)`, and the same but with `.not_change` for the invalid case

File: app/controllers/v1/sessions_controller.rb (hunk 17 above)
Comment (ivanfelfo) [reply app/controllers/v1/sessions_controller.rb]: it's working

File: spec/requests/sessions/create_spec.rb (hunk 21 above)
Comment (ivanfelfo) [reply spec/requests/sessions/create_spec.rb:8]: 👍

File: spec/requests/sessions/create_spec.rb (hunk 22 above)
Comment (ivanfelfo) [reply spec/requests/sessions/create_spec.rb:12]: 👍

File: spec/requests/sessions/create_spec.rb (hunk 23 above)
Comment (ivanfelfo) [reply spec/requests/sessions/create_spec.rb:23]: 👍

File: spec/requests/users/create_spec.rb (hunk 25 above)
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:7]: yes, it'll be my next PR

File: spec/requests/users/create_spec.rb (hunk 24 above)
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:10]: could it be 'creates and saves user in db' ??

File: spec/requests/users/create_spec.rb (hunk 26 above)
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:16]: 👍

File: spec/requests/sessions/create_spec.rb [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+        expect(response).to have_http_status(401)
+      end
+    end
+
+    context 'when leaving a blank parameter' do
+      let(:params) { { user: { email: 'mail@mail.com' } } }
+
+      #   it  'responds with code 401' do
+      #     expect(response).to have_http_status(401)
+      #   end /// it is returning nil
```
Comment (martinjaimem) [inline spec/requests/sessions/create_spec.rb:37]: We should fix this

File: spec/requests/users/create_spec.rb [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+        subject
+        expect(User.count).to eq(1)
+      end
+
+      it 'responds with 400' do
+        subject
+        expect(response).to have_http_status(400)
+      end
+
+      it 'will NOT change the user count' do
```
Comment (martinjaimem) [inline spec/requests/users/create_spec.rb:43]:
```suggestion
      it 'doesn\'t change the user count' do
```

File: spec/requests/users/create_spec.rb [hunk 29]
```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'POST /v1/users', type: :request do
+  subject { post v1_user_registration_path, params: params, as: :json }
+  let(:params) { { user: { email: 'mail@mail.com', password: 'hello123', gender: true } } }
+
+  context 'when valid' do
+    it 'creates and saves user in db' do
```
Comment (martinjaimem) [inline spec/requests/users/create_spec.rb:10]:
```suggestion
    it 'creates a user in db' do
```

File: spec/requests/users/create_spec.rb [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+
+  context 'when not valid' do
+    context 'when trying to register a user with an email that already exists' do
+      let!(:created_user) do
+        User.create(email: 'mail@mail.com', password: 'hello123', gender: true)
+      end
+
+      it 'does not create a new user' do
+        subject
+        expect(User.count).to eq(1)
```
Comment (martinjaimem) [inline spec/requests/users/create_spec.rb:35]:
```suggestion
        expect { subject }.not_to change(User, :count)
```

File: spec/requests/users/create_spec.rb (hunk 28 above)
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:43]: 👍

File: spec/requests/users/create_spec.rb (hunk 29 above)
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:10]: 👍

File: spec/requests/users/create_spec.rb (hunk 30 above)
Comment (ivanfelfo) [reply spec/requests/users/create_spec.rb:35]: I have the same example a couple of lines below, so I'm deleting this one.

File: spec/requests/sessions/create_spec.rb (hunk 27 above)
Comment (ivanfelfo) [reply spec/requests/sessions/create_spec.rb]: resolved

**Review by juanmferreira93 — APPROVED** [review]

**Comment by juanmferreira93** [comment]
Great work! 🚀

**Review by martinjaimem — APPROVED** [review]
